contracts/perun-vchannel-typescript/src/entry.rs

```diff
@@ -0,0 +1,1633 @@
+    channel_constants: &ChannelConstants,
+) -> Result<(), Error> {
+    const FUNDER_INDEX: usize = 0;
+
+    debug!("check_valid_start");
+
+    // Upon start of a channel, the channel constants are stored in the args field of the pcts output.
+    // We uniquely identify a channel through the combination of the channel id (hash of ChannelParameters,
+    // which is part of the ChannelConstants) and the "thread token".
+    // The thread token contains an OutPoint and the channel type script verifies, that that outpoint is
+    // consumed in the inputs of the transaction that starts the channel.
+    // This means: Once a (pcts-hash, channel-id, thread-token) tuple appears once on chain and is recognized
+    // as the on-chain representation of this channel by all peers, no other "copy" or "fake" of that channel
+    // can be created on chain, as an OutPoint can only be consumed once.
+
+    // here, we verify that the OutPoint in the thread token is actually consumed.
+    verify_thread_token_integrity(&channel_constants.thread_token())?;
+    debug!("verify_thread_token_integrity passed");
+
+    // We verify that the channel id is the hash of the channel parameters.
+
+    let vc_chanid = new_status.vcstate().channel_id();
+
+    verify_vchannel_id_integrity(&vc_chanid, &channel_constants.params())?;
+    debug!("verify_channel_id_integrity passed");
+
+    // We verify that the pcts is guarded by the pcls script specified in the channel constants
+    verify_valid_lock_script(channel_constants)?;
+    debug!("verify_valid_lock_script passed");
+
+    // We verify that the channel participants have different payment addresses
+    // For this purpose we consider a payment address to be the script hash of the lock script used for payments to that party
+    verify_different_payment_addresses(channel_constants)?;
+    debug!("verify_different_payment_addresses passed");
+
+    // We verify that there are no funds locked by the pfls hash of this channel in the inputs of the transaction.
+    // This check is not strictly necessary for the current implementation of the pfls, but it is good practice to
+    // verify this anyway, as there is no reason to include funds locked for any channel in the input of a transaction
+    // that creates a new channel besides trying some kind of attack.
+    verify_no_funds_in_inputs(channel_constants)?;
+    debug!("verify_no_funds_in_inputs passed");
+
+    // We verify that the state the channel starts with is valid according to the utxo-adaption of the perun protocol.
+    // For example, the channel must not be final and the version number must be 0.
+    verify_vcstate_valid_as_start(
+        &new_status.vcstate(),
+        channel_constants.pfls_min_capacity().unpack(),
+    )?;
+    debug!("verify_vcstate_valid_as_start passed");
+
+    // Here we verify that the first party completes its funding and that itsfunds are actually locked to the pfls with correct args.
+    verify_funding_in_outputs(
```

Comment:
Again, Virtual channels are not funded...
So there should not be funds in outputs